src/plans/vectorial_plan.jl

```diff
@@ -0,0 +1,478 @@
+that if this is expensive even to compute a single component, all of `f` has to be evaluated
+
+For the [`ComponentVectorialType`](@ref) of `f`, each of the component functions
+is a (classical) objective and hence has a gradient ``\oepratorname{grad} f_i(p) ∈ T_p\mathcal M``.
+Putting these gradients into a vector the same way as the functions, we obtain a
+[`ComponentVectorialType`](@ref)
+
+```math
+\operatorname{grad} f(p) = \Bigl( \operatorname{grad} f_1(p), \operatorname{grad} f_2(p), …, \operatorname{grad} f_n(p) \Bigr)^{\mathrm{T}}
+∈ (T_p\mathcal M)^n
+```
+
+And advantage here is, that again the single components can be evaluated individually
+
+# Fields
+
+* `costs!!`:          the cost function ``f``, which can take different formats
+* `cost_type`:     indicating / string data for the type of `f`
+* `jacobian!!:     the jacobian of ``f``
+* `jacobian_type`: indicating / storing data for the type of ``J_f``
+* `parameters`:    the nunmber `n` from above, that is the size of the vector ``f`` returns.
+
+# Constructor
+
+VectorGradientFunction(f, Jf, range_dimension;
+    evaluation::AbstractEvaluationType=AllocatingEvaluation(),
+    function_type::AbstractVectorialType=FunctionVectorialType(),
+    jacobian_type::AbstractVectorialType=FunctionVectorialType(),
+)
+"""
+struct VectorGradientFunction{
+    E<:AbstractEvaluationType,
+    FT<:AbstractVectorialType,
+    JT<:AbstractVectorialType,
+    F,
+    J,
+    I<:Integer,
+} <: Function
+    costs!!::F
+    cost_type::FT
+    jacobian!!::J
+    jacobian_type::JT
+    range_dimension::I
+end
+
+function VectorGradientFunction(
+    f::F,
+    Jf::J,
+    range_dimension::I;
+    evaluation::E=AllocatingEvaluation(),
+    function_type::FT=FunctionVectorialType(),
+    jacobian_type::JT=FunctionVectorialType(),
+) where {
+    I,F,J,E<:AbstractEvaluationType,FT<:AbstractVectorialType,JT<:AbstractVectorialType
+}
+    return VectorGradientFunction{E,FT,JT,F,J,I}(
+        f, function_type, Jf, jacobian_type, range_dimension
+    )
+end
+
+@doc raw"""
+    get_cost(M::AbstractManifold, vgf::VectorGradientFunction, p, i, range=nothing)
+
+Evaluate the ``i``th component or components of the [`VectorGradientFunction`](@ref)
+cost at `p`. The `range` can be used to speficy a potential range.
+This is currently ignored, but kept for consistency with the gradient acces functions.
+
+Since `i` is assumed to be a linear index, you can provide
+
+* a single integer
+* a `UnitRange` to specify a range to be returned like `1:3`
+* a `BitVector` specifying a selection
+* a `AbstractVector{<:Integer}` to specify indices
+* `:` to return the vector of all gradients
+
+"""
+get_cost(M::AbstractManifold, vgf::VectorGradientFunction, p, i, range=nothing)
+function get_cost(
+    M::AbstractManifold,
+    vgf::VectorGradientFunction{<:AllocatingEvaluation,<:FunctionVectorialType},
+    p,
+    i,
+    range=nothing,
+)
+    return vgf.costs!!(M, p)[i]
```

Comment:
Why isn't this `vgf.costs!!(M, p, i)`?

Reply:
The cost returns a vector.
This is the case (`FunctionVectorialType`)  where we have a single cost that returns a vector. So to access the I-th element of a vector after a function call, this is correct
(but yes with these different cases we have now, one has to bet always careful with these, sure)

Reply:
Hm, OK, let's not touch this for now.

Reply:
What would you want to change here? Wrap the cost as well in something?
I do agree that we should avoid a Vector{Function}, but we never explicitly need such a type, since we have to also distinguish ranges there. So I think there is no need to wrap the cost and implement the function you mentioned here)

Reply:
I was thinking of a case where the user might want to optimize evaluation of a subset of constraints, i.e. get the benefits of both "vector functions" and "function returning a vector" approaches. Then they would need to get the indices, but such case might be too niche to care about at this point.

Reply:
I fear that the current code is already more clever than I am, I am fighting quite quite a bit to get this working at all.

Your proposal sounds fair (and might even be reasonable for the gradients), but I am not clever enough for that I fear.  This will take a few weeks to work anyways in this simpler scenario is what I estimate, since the current different variants make all this quite tricky internally.